packages/x-data-grid-pro/src/DataGridPro/DataGridPro.tsx

```diff
@@ -483,6 +483,18 @@ DataGridProRaw.propTypes = {
    * @default false
    */
   keepNonExistentRowsSelected: PropTypes.bool,
+  /**
+   * Used together with `unstable_dataSource` to enable lazy loading.
+   * If enabled, the grid stops adding `paginationModel` to the data requests (`getRows`)
+   * and starts sending `start` and `end` values depending on the loading mode and the scroll position.
+   * @default false
+   */
+  lazyLoading: PropTypes.bool,
```

Comment:
I advocated to move it to root props because:
1. Other feature flags like `treeData` are in root props
2. We should not pollute the `dataSource` interface with the frontend / Data Grid specific things.

As I understand, the `dataSource` should only contain information about the actual data source (backend which provides the data), like how to interact with it and extract the data from it.

On a rethought, I tend to believe that `lazyLoading` could also explain the nature of data a particular backend provides, so it kind-of makes sense to make it part of the dataSource (But in that case, should we make other feature flags like `treeData`, `rowGrouping`, `aggregation`, etc. part of it too?). About `lazyLoadingRequestThrottleMs`, it is a frontend specific thing and doesn't have to do anything with the actual data source. Should we make it part of the data source too?

<hr />

I think we should take a step back and agree a bit what we actually mean (and want our users to understand) by the term `dataSource`.

As per the definition in our [docs](https://mui.com/x/react-data-grid/server-side-data/#data-source):

> It's an abstraction layer between the Data Grid and the server, providing a simple interface **for interacting with the server.** Think of it like an intermediary handling the communication between the Data Grid (client) and the actual data source (server).

What I understand by this is, it's a way for the users to explain to the Data Grid the shape of the backend server which the Data Grid could use to extract a certain data chunk.

The behavior of the Data Grid itself (like how much time it should throttle `lazyLoadingRequestThrottleMs`, etc) seem to be best suited in Data Grid props. One example is where making it part of data source would cause unexpected outcome it when the users want to update the throttle milliseconds. (The data source object would be updated and the Data Grid would reset the cache and datasource data which is not desired)

So we could agree on either of the following and explain in the docs clearly:
1. `dataSource` object contains static information (or metadata) about the actual backend server, like:
	- `getRows` that provides a way to get rows from the backend
	- `updateRows` that provides a way to update the rows on the backend
	- Resolver methods that allow the Data Grid to extract relevant information from the backend data (`getGroupKeys`, `getChildrenCount`, etc.)
2. `dataSource` object encapsulates static information about the actual data source (point 1) plus the (frontend) parameters that correspond to how the Data Grid should interact with the data source (like `lazyLoadingRequestThrottleMs`).

If we go with 2, we have to make sure the data source doesn't reset when any of the frontend params update but does reset when any of the backend params change.

Which one makes more sense to you?

1 = 👍 
2 = 🚀

Reply:
I am more for keeping it generic
even more generic than what it is now. like not using "rows" but "items" for the records since there is a potential to share this part with other X component that will load server data
This is now going into a discussion outside of the scope of PR, but `getGroupKeys`, `getChildrenCount` and similar might be merged into something like `getMetaData` where server has to provide a specific interface based on the component or the UI it is filling in

Reply:
Sorry, I lost track of this thread.

> One example is where making it part of data source would cause unexpected outcome it when the users want to update the throttle milliseconds. (The data source object would be updated and the Data Grid would reset the cache and datasource data which is not desired)

I think this is the most convincing argument 👍🏻

Reply:
>  like not using "rows" but "items" for the records since there is a potential to share this part with other X component that will load server data

Here was the argument to name it `rows` when we discussed making it generic. Do you think it's convincing enough?

> getGroupKeys, getChildrenCount and similar might be merged into something like getMetaData where server has to provide a specific interface based on the component or the UI it is filling in

I'm not completely sure about this, we already [discussed](https://github.com/mui/mui-x/pull/12317#discussion_r1616959519) to make them Grid props instead, I still feel the best would be to have the data source interface specific to the Data Grid and share the concept instead. Here's the pseudo code shared by Andrew after a meeting on the same topic: https://www.notion.so/mui-org/mui-x-Data-provider-discussion-2024-02-09-0445da5318a349c3aa28259553cbecda?pvs=4#fcccad9ddb934c6fb07e895e0704bc88